### PR TITLE
refactor: clean up `errors.rs` and `error_codes_check.rs`

### PR DESCRIPTION
**Pull request by Ezrashaw**
`errors.rs` is basically unused now, `error_codes_check.rs` is useful but not well commented, etc. It also doesn't check certain things which are certainly not correct. For example, `E0505` has a UI test in `src/test/ui/error-codes/` but that test actually outputs `E0504`?! Other issues like these exist. I've implemented these with "warnings" which are a bit rough around the edges but should be removed eventually.

r? @GuillaumeGomez (again not sure if you want to review but its relevant to you)

### CODE REVIEW

**Comment by Ezrashaw** [comment]
The CI is failing because of a check ensuring error code explanations are not removed. I'm not a big fan of this, what help can outdated documentation be? I want guidance on this and then I'll remove that CI check.

File: compiler/rustc_error_codes/src/error_codes/E0729.md [hunk 1]
```diff
@@ -1,30 +0,0 @@
-Support for Non-Lexical Lifetimes (NLL) has been included in the Rust compiler
```
Comment (GuillaumeGomez) [inline compiler/rustc_error_codes/src/error_codes/E0729.md:1]: Please revert this change. Error code explanations should never be removed. Please only update it to mention that it's not emitted by the compiler anymore.

**Comment by GuillaumeGomez** [comment]
This is a really nice improvements, thanks a lot for working on this. Once my comment is fixed, it'll be good to go!

File: src/tools/tidy/src/error_codes.rs [hunk 2]
```diff
@@ -0,0 +1,397 @@
...
+    errors: &mut Vec<String>,
+    no_longer_emitted: &[String],
+) {
+    // We want error codes which match the following cases:
+    //
+    // * foo(a, E0111, a)
+    // * foo(a, E0111)
+    // * foo(E0111, a)
+    // * #[error = "E0111"]
+    let regex = Regex::new(r#"[(,"\s](E\d{4})[,)"]"#).unwrap();
```
Comment (klensy) [inline src/tools/tidy/src/error_codes.rs:345]: regex compile can be moved to lazy_static/once_cell?

File: src/tools/tidy/src/error_codes.rs [hunk 3]
```diff
@@ -0,0 +1,397 @@
...
+        );
+    }
+
+    no_longer_emitted_codes
+}
+
+/// This function returns a tuple indicating whether the provided explanation:
+/// a) has a code example, tested or not.
+/// b) has a valid doctest
+fn check_explanation_has_doctest(explanation: &str, err_code: &str) -> (bool, bool, bool, bool) {
```
Comment (mejrs) [inline src/tools/tidy/src/error_codes.rs:232]: Functions like this that return a bunch of booleans are error prone and not self-documenting. 

An alternative would be to return a struct, or return a tuple of boolean-like enums, e.g. 
```rust
enum HasDocTest{
    Yes,
    No
}

enum HasUiTest{
    Yes,
    No
}
// etc
```

or:

```rust
struct TestChecks{
    has_ui: bool,
    has_doc: bool,
    // ...
}
```

File: src/tools/tidy/src/error_codes.rs (hunk 2 above)
Comment (mejrs) [reply src/tools/tidy/src/error_codes.rs:345]: It seems like it's compiled only once, that shouldn't matter.

File: src/tools/tidy/src/error_codes.rs [hunk 4]
```diff
@@ -0,0 +1,397 @@
...
+    // Print any errors.
+    for error in errors {
+        tidy_error!(bad, "{}", error);
+    }
+}
+
+/// Stage 1: Parses a list of error codes from `error_codes.rs`.
+fn extract_error_codes(root_path: &Path, errors: &mut Vec<String>) -> Vec<String> {
+    let file = fs::read_to_string(root_path.join(Path::new(ERROR_CODES_PATH)))
+        .unwrap_or_else(|e| panic!("failed to read `error_codes.rs`: {e}"));
```
Comment (mejrs) [inline src/tools/tidy/src/error_codes.rs:65]: It's nice when errors like this one post the actual path, so that I don't have to end up debugging it.

```suggestion
    let path = root_path.join(Path::new(ERROR_CODES_PATH));
    let file = fs::read_to_string(&path)
        .unwrap_or_else(|e| panic!("failed to read `{path:?}`: {e}"));
```

(Same for some other panics/messages in here)

File: src/tools/tidy/src/error_codes.rs [hunk 5]
```diff
@@ -0,0 +1,397 @@
...
+
+    (found_code_example, found_proper_doctest, emit_ignore_warning, emit_no_longer_warning)
+}
+
+// Stage 3: Checks that each error code has a UI test in the correct directory
+fn check_error_codes_tests(root_path: &Path, error_codes: &[String], errors: &mut Vec<String>) {
+    let tests_path = root_path.join(Path::new(ERROR_TESTS_PATH));
+
+    // Some warning counters, this whole thing is clunky but'll be removed eventually.
+    let mut no_ui_test = 0;
```
Comment (mejrs) [inline src/tools/tidy/src/error_codes.rs:271]: Instead of a counter, why not have a vec of strings like you do elsewhere, and just print them if they're not empty? Otherwise this is just going to print that a bunch of these codes need tests, but not *which ones*.

**Review by mejrs — COMMENTED** [review]
I think the error messages are generally good 👍 I have some comments though

File: src/tools/tidy/src/error_codes.rs (hunk 2 above)
Comment (Ezrashaw) [reply src/tools/tidy/src/error_codes.rs:345]: It is, this was a change from how it was previously done I believe.

File: src/tools/tidy/src/error_codes.rs (hunk 5 above)
Comment (Ezrashaw) [reply src/tools/tidy/src/error_codes.rs:271]: I did this originally, and I agree that it's simpler. However with the new warnings I've introduced, there are ~250 warnings that would be emitted. The warnings aren't really permanent and are more just tracking progress. When these are fixed (I'll have some follow up PRs), they'll become hard errors, which simplifies the code a lot.

File: src/tools/tidy/src/error_codes.rs (hunk 3 above)
Comment (Ezrashaw) [reply src/tools/tidy/src/error_codes.rs:232]: As below, I can change this if you want, but these aren't permanent and just exist until I fix up the warnings and remove half of that function.

File: compiler/rustc_error_codes/src/error_codes/E0729.md (hunk 1 above)
Comment (Ezrashaw) [reply compiler/rustc_error_codes/src/error_codes/E0729.md:1]: Done. I don't really like this, what use are these old docs (especially considering lots aren't documented)? If your using an old compiler, you can `rustc --explain Exxxx` and if you are using an up-to-date compiler then there is literally no reason you could ever need those docs.

File: src/tools/tidy/src/error_codes.rs (hunk 4 above)
Comment (Ezrashaw) [reply src/tools/tidy/src/error_codes.rs]: Think I got them all.

**Comment by mejrs** [comment]
I checked it out locally and `tidy` spits out the following: 
```
WARNING: 10 error codes are undocumented. This *will* become a hard error.
Found 505 error codes
WARNING: 36 error codes use the ignore header. This should not be used, add the error codes to the `IGNORE_DOCTEST_CHECK` constant instead. This *will* become a hard error.
WARNING: 36 error codes don't have a code example, all error codes are expected to have one (even if untested). This *will* become a hard error.
WARNING: 68 error codes are no longer emitted and should be removed entirely. This *will* become a hard error.
WARNING: 8 error codes have a UI test file, but don't contain their own error code!
WARNING: 272 error codes need to have at least one UI test in the `src/test/ui/error-codes/` directory`! This *will* become a hard error.
WARNING: 6 error codes are used when they are marked as "no longer emitted"
* 392 features
fmt check
Build completed successfully in 0:01:09
```

We shouldn't emit warnings like that unless it's the user's fault. There should be no "WARNING" in all caps and no "This *will* become a hard error."

I think it would be fine to have something like this as an opt-in (that would definitely be helpful for your followup PRs and anyone helping), but not by default.

**Comment by Ezrashaw** [comment]
> I think it would be fine to have something like this as an opt-in (that would definitely be helpful for your followup PRs and anyone helping), but not by default.

Could that be hardcoded? I'll also fix up all poor counter logic and stuff, we can just print 100s of messages if its opt-in.

**Comment by mejrs** [comment]
> Nevermind, we can use `--verbose`

Or you could make a `--pedantic` setting. It's generally understood that if you enable something called pedantic that the errors may be terse.

**Comment by Ezrashaw** [comment]
`tidy` already has a bool for verbose in `main.rs` so I'll just use that. Pedantic implies failing the build but this is just supplementary information.

File: src/tools/tidy/src/error_codes.rs (hunk 3 above)
Comment (Ezrashaw) [reply src/tools/tidy/src/error_codes.rs:224]: Somewhat fixed with the `--verbose` changes and code simplification in that regard. This will be further improved soon when these warnings are removed.

File: src/tools/tidy/src/error_codes.rs [hunk 6]
```diff
@@ -0,0 +1,381 @@
...
+                "Found valid file `{}` in error code docs directory without corresponding \
+                entry in `error_code.rs`",
+                path.display()
+            ));
+            return;
+        }
+
+        // `has_test.0` checks whether the error code has any (potentially untested) code example.
+        // `has_test.1` checks whether the error code has a proper (definitely tested) doctest.
+        let has_test = check_explanation_has_doctest(&contents, &err_code);
```
Comment (mejrs) [inline src/tools/tidy/src/error_codes.rs:178]: If you don't want to change this function now, please destructure the tuple:
```suggestion
        let (found_code_example, found_proper_doctest, emit_ignore_warning, emit_no_longer_warning) = check_explanation_has_doctest(&contents, &err_code);
```
(This also has the advantage of being able to express what they are in *code* rather than in *comments*)

**Review by mejrs — APPROVED** [review]
It looks good, thanks for making the changes, I just have one final nit.

File: compiler/rustc_error_codes/src/error_codes/E0729.md (hunk 1 above)
Comment (GuillaumeGomez) [reply compiler/rustc_error_codes/src/error_codes/E0729.md:1]: Because people might still use old rustc version and look for their explanation online. If we remove them, they disappear in latest docs version, forcing you to look into the older ones (which aren't very easy to find if you don't know where to look).

File: src/tools/tidy/src/errors.rs [hunk 7]
```diff
@@ -1,77 +0,0 @@
...
-
-        tidy_error!(bad, "duplicate error code: {}", code);
-        for &(ref file, line_num, ref line) in entries.iter() {
-            tidy_error!(bad, "{}:{}: {}", file.display(), line_num, line);
-        }
-    }
-
-    if !*bad {
-        println!("* {} error codes", map.len());
-        println!("* highest error code: E{:04}", max);
```
Comment (GuillaumeGomez) [inline src/tools/tidy/src/errors.rs:75]: Just realized but this file is still in use for at least this part.

Comment (Ezrashaw) [reply src/tools/tidy/src/errors.rs:75]: I had already copied over the "error code count". Just added the "highest error code".

File: compiler/rustc_error_codes/src/error_codes/E0729.md (hunk 1 above)
Comment (Ezrashaw) [reply compiler/rustc_error_codes/src/error_codes/E0729.md:1]: Do people ever use `--explain` or is that just me lol? It's very interesting, error codes aren't gated by edition. *Personally*, I would prefer purging all the old ones and adding a note like, "If an error code is not listed, then it has been removed in a recent `rustc` version, please update. You can find old error code documentation \[here](link) or by using the `--explain` flag." I thought that you're on your own if you don't update because of the backwards-compatibility promises.

Comment (GuillaumeGomez) [reply compiler/rustc_error_codes/src/error_codes/E0729.md:1]: I meant online. Here to be exact: https://doc.rust-lang.org/error_codes/error-index.html

Comment (Ezrashaw) [reply compiler/rustc_error_codes/src/error_codes/E0729.md:1]: Yeah, I know. But we don't have to support them? For example, they don't get bug fixes. Why should they get docs?

Comment (GuillaumeGomez) [reply compiler/rustc_error_codes/src/error_codes/E0729.md:1]: Error codes aren't *bugs*, they're errors in user's code.

Comment (Ezrashaw) [reply compiler/rustc_error_codes/src/error_codes/E0729.md:1]: What I'm saying (rather badly) is that we (all the rustc developers) do not support versions of rustc that are not up to date. If borrow checker bug is found in rustc for an old rustc version (say 1.60) then *it isn't fixed*. The fix for people who use affected versions of rustc *is to update*, this works because Rust is always backwards-compatible (there is no effort to use a newer rustc). However the "problem" I have is that this isn't true for long-form error code docs, we keep the old docs even though *users should update*. Another example: in stdlib, we don't keep docs for removed functions (however rare that is).

Comment (GuillaumeGomez) [reply compiler/rustc_error_codes/src/error_codes/E0729.md:1]: We never removed a function in stdlib because it would break semver.

Also I think we're not talking about the same thing at all. Just in case: imagine a company using only rust 1.10 for whatever reason. They have a very old code and therefore see completely legit old error codes thrown out by the compiler. They can look it up locally of course, but let's say they want to ask for extra explanations about the explanation and therefore share its online location. However, since the error code is long time since then, it was removed. Then there is no way to look for help for the user.

It's somewhat similar to why we don't re-use error codes: because we want to confuse users using old error codes which would have invalid explanations.

So, the error code and its explanations are still perfectly valid for the context in which they were written. If someone is looking for its explanation, it means it has a use for it. And once again, an error code doesn't mean something is wrong with the compiler, it means it's a legit error from the user's code.

Final point: it's current in the industry to pinpoint a specific compiler version and use it for a few years before migrating to a new one. For such users, the documentation needs to remain available.

Comment (Ezrashaw) [reply compiler/rustc_error_codes/src/error_codes/E0729.md:1]: > We never removed a function in stdlib because it would break semver.

What about editions?

Fair points, I don't entirely agree but it makes sense. 

As an aside, would you ever write documentation for error codes that are no longer emitted? Also, is this resolved then?

Comment (mejrs) [reply compiler/rustc_error_codes/src/error_codes/E0729.md:1]: > What about editions?

That cannot be done across editions. It is a very limited mechanism.

Either way, the policy is that old error codes are not removed from the index, and this isn't the right place to discuss changing the policy 😛 Consider opening a zulip thread or an issue if you feel strongly about it - but I think you'll find many share Guillaume's stance.

Comment (GuillaumeGomez) [reply compiler/rustc_error_codes/src/error_codes/E0729.md:1]: > What about editions?

We never removed std libs API in editions. Deprecated for sure but removed, never.

> As an aside, would you ever write documentation for error codes that are no longer emitted?

The only goal here is to prevent to remove knowledge. Adding content for outdated errors is another matter. So the answer is no, I would not write documented for removed error codes.

> Also, is this resolved then?

You un-removed the file so all good for me.

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
Looks good to me, thanks for working on this! Do you see anything else @mejrs and @klensy ?

**Comment by mejrs** [comment]
> Looks good to me, thanks for working on this! Do you see anything else @mejrs and @klensy ?

I'm happy with it :)

**Comment by GuillaumeGomez** [comment]
Then let's go!

@bors r=mejrs,klensy,GuillaumeGomez rollup
